socs/agents/hwp_pid/agent.py

```diff
@@ -27,88 +84,116 @@ def __init__(self, agent, ip, port, verbosity):
+            The most recent data collected is stored in the session data in the
+            structure::
+
+                >>> response.session['data']
+                {'current_freq': 0,
+                 'target_freq': 0,
+                 'direction': 1,
+                 'last_updated': 1649085992.719602}
         """
-        if self._initialized and not params['force']:
-            self.log.info("Connection already initialized. Returning...")
-            return True, "Connection already initialized"
+        pid = pd.PID(ip=self.ip, port=self.port, verb=self._verbosity)
+        self.log.info("Connected to PID controller")
 
-        with self.lock.acquire_timeout(10, job='init_connection') as acquired:
-            if not acquired:
-                self.log.warn(
-                    'Could not run init_connection because {} is already running'.format(self.lock.job))
-                return False, 'Could not acquire lock'
+        self._clear_queue()
 
-            try:
-                self.pid = pd.PID(ip=self.ip, port=self.port,
-                                  verb=self._verbosity)
-                self.log.info('Connected to PID controller')
-            except BrokenPipeError:
-                self.log.error('Could not establish connection to PID controller')
-                reactor.callFromThread(reactor.stop)
-                return False, 'Unable to connect to PID controller'
+        sample_period = 5.0
+        last_sample = 0.0
+        session.set_status("running")
+        while session.status in ["starting", "running"]:
+            now = time.time()
+            if now - last_sample > sample_period:
+                self._get_data_and_publish(pid, session)
+                last_sample = now
 
-        self._initialized = True
+            self._process_actions(pid)
+            time.sleep(0.2)
 
-        # Start 'acq' Process if requested
-        if params['auto_acquire']:
-            self.agent.start('acq')
+        return True, "Exited main process"
 
-        return True, 'Connection to PID controller established'
+    def _main_stop(self, session, params):
+        """Stop main process"""
+        session.set_status("stopping")
+        return True, "Set main status to stopping"
 
+    @defer.inlineCallbacks
     def tune_stop(self, session, params):
         """tune_stop()
 
         **Task** - Reverse the drive direction of the PID controller and
         optimize the PID parameters for deceleration.
 
         """
-        with self.lock.acquire_timeout(3, job='tune_stop') as acquired:
-            if not acquired:
-                self.log.warn(
-                    'Could not tune stop because {} is already running'.format(self.lock.job))
-                return False, 'Could not acquire lock'
-
-            self.pid.tune_stop()
-
-        return True, 'Reversing Direction'
+        action = Actions.TuneStop(**params)
+        self.action_queue.put(action)
+        session.data = yield action.deferred
```

Comment:
Reminder to keep session.data as a dict.  Maybe standardize on `session.data['output'] = ...` or something like that?

Also -- what happens here if there's an errback?  Do you need to catch that?

Reply:
If there's an errback, the `yield` will just throw the error. I find this to be the easiest way to find and debug errors in ocs operations anyway, so I don't see the need to catch and handle manually.